docs/src/proposals/stake-v2.md

```diff
@@ -0,0 +1,215 @@
+if the size is `200`, and a `StakeV2` if the size is `256`.
+
+### Epoch Boundary Calculations
+
+While we're at it, new stake accounts can also solve point number 2, by having
+their rewards paid through an update instruction, and not automatically at every
+epoch boundary.
+
+During epoch boundary, the validator vote accounts receive all stake V2
+account rewards as withheld lamports in their account. They cannot
+touch these, and stake accounts may claim them.
+
+The biggest users of stake accounts, ie stake pools and stake bots, already
+touch stake accounts every epoch during their bookkeeping, so this adds trivial
+work for them.
+
+Legacy stake accounts are unaffected, and inflation continues to pay out their rewards
+at epoch boundaries automatically.
+
+## New Stake Program Instructions
+
+`Split` / `Merge` / `Deactivate` are only allowed if the stake account has
+been updated this epoch, or after deactivation epoch if applicable.
+
+Side note: deactivate might actually be OK, since we could technically calculate
+the rewards even after deactivation.
+
+### UpgradeToV2
+
+Given a V1 stake account, a signature from the stake or withdraw authority, and
+a funding account, reallocate the existing stake account and convert it to a V2
+stake. The funding account adds the lamports necessary to cover the new rent-exemption.
+
+Side note: this would also be a great time to include the rent-exemption in the
+delegation. That way, smaller stakers get a tiny bit more, and there's a couple of other
+nice side effects. Ask me about it.
+
+### Redelegate
+
+Given a V2 stake account, vote account, and stake authority signature, change the
+redelegation. Follows most of the same rules as delegate.
+
+If it's going through a multi-epoch redelegation, be sure that previous epoch
+rewards have been claimed, and update the vote pubkey / credits / epoch.
+
+### (Also in the Vote Program) ClaimRewards
+
+Given only a V2 stake account and the delegated vote account (or two vote accounts
+during multi-epoch redelegation), calculate the rewards gained since the last
+claim, and withdraw the lamports from the vote account.
+
+This instruction is permissionless.
+
+For easier use with existing tools, if a redelegation is complete, move the
+redelegation information over into the main delegation.
+
+Implementation note: claiming lamports from the vote account and also updating the
+stake account is tricky since only the owning program can deduct lamports or modify
+data, and we need to do both conditionally.
+
+Potential solution: the vote program CPIs into the stake program to update the
+credits observed, then transfers the lamports from the vote account to the stake account.
+The stake program must enforce that only the vote program can call it, otherwise
+it's possible to deprive an account of their rewards.
+
+BIG ISSUE TO CONSIDER: vote accounts only store 64 epochs of credits, which means
+that we can only accurately calculate 64 epochs of credits.
```

Comment:
Not a deal-breaker: I bet it would be appreciated by US folks if they could wait for over 1 year before collecting rewards (due to US tax laws).

How (im)possible is it to change vote accounts to store more epochs?

Reply:
ha funny. that's assuming the cap gain clock would even start in the same way after such a change, pretty sure it wouldn't but ianal and certainly not a tax one.

> How (im)possible is it to change vote accounts to store more epochs?

Not impossible.  Just need to realloc every vote account in the system for more room, and pay the extra rent.  So, practically impossible for something that likely doesn't help anyway :)

Reply:
This does raise a good point though.  Adding optionality may change the asset class in some jurisdictions.  Specifically in the US this may qualify as "constructed receipt"

Reply:
tfw you need legal to review your PR :sweat_smile: